cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java

```diff
@@ -1675,13 +1669,72 @@ private String generatePerBrokerConfiguration(NodeRef node, KafkaPool pool, Map<
+     * | PRE_MIGRATION  | KRaft        | Zk           |
+     * |                | Zk           |              | ---> controllers deployed
+     * |                | Zk-migration |              |
+     * +----------------+--------------+--------------+
+     * | MIGRATION      | KRaft        | KRaft        |
+     * |                | Zk           | Zk           | ---> brokers rolled
+     * |                | Zk-migration | Zk-migration |
+     * +----------------+--------------+--------------+ ---> strimzi.io/kraft = enabled
+     * | POST_MIGRATION | KRaft        | KRaft        |
+     * |                | Zk           |              | ---> brokers rolled
+     * |                | Zk-migration |              |
+     * +----------------+--------------+--------------+
+     * | KRAFT          | KRaft        | KRaft        | ---> controllers rolled
+     * +----------------+--------------+--------------+
+     *
+     * @param node node on which the configuration is applied
+     * @param pool pool hosting the node
+     * @param builder KafkaBrokerConfigurationBuilder instance to use to build the node configuration
+     */
+    private void buildBrokerConfiguration(NodeRef node, KafkaPool pool, KafkaBrokerConfigurationBuilder builder) {
+        if (node.broker()) {
+            if (this.kafkaMetadataConfigState.isZooKeeperOrMigration()) {
+                builder.withZookeeper(cluster);
+                LOGGER.infoCr(reconciliation, "withZookeeper on node [{}]", node.podName());
+            }
+
+            if (this.kafkaMetadataConfigState.isMigration()) {
+                builder.withZooKeeperMigration(true);
+                LOGGER.infoCr(reconciliation, "withZooKeeperMigration on node [{}]", node.podName());
+            }
+
+            if (this.kafkaMetadataConfigState.isMigrationOrKRaft()) {
+                builder.withKRaft(cluster, namespace, pool.processRoles, nodes());
+                LOGGER.infoCr(reconciliation, "withKRaft on node [{}]", node.podName());
+            }
+        }
+
+        if (node.controller()) {
+            if (this.kafkaMetadataConfigState.isPreMigrationOrKRaft() && this.kafkaMetadataConfigState.isZooKeeperOrPostMigration()) {
+                builder.withZookeeper(cluster);
+                builder.withZooKeeperMigration(true);
+                LOGGER.infoCr(reconciliation, "withZookeeper on node [{}]", node.podName());
+                LOGGER.infoCr(reconciliation, "withZooKeeperMigration on node [{}]", node.podName());
+            }
+
+            // additional check not having broker role as well because in mixed mode
+            // (only in KRaft state, migration not supported) it could end to duplicate the configuration
+            if (this.kafkaMetadataConfigState.isPreMigrationOrKRaft() && !node.broker()) {
+                builder.withKRaft(cluster, namespace, pool.processRoles, nodes());
+                LOGGER.infoCr(reconciliation, "withKRaft on node [{}]", node.podName());
+            }
```

Comment:
* This looks super strange. You do not use if-else but use two independent ifs instead. So as a result, you call many methods multiple times under some conditions. I think it would be great if we could do it in a more clear and readable way with if-else-if-else branching.
* TBH, even the split between generatePerBrokerConfiguration and buildBrokerConfiguration methods seems quite confusing. I wonder if more code duplication would give us more readable and maintainable code.
* The INFO level log messages seem to be way too aggressive? Are they some debugging leftover? Should we remove them?

Reply:
So ....
* I think this was the first method I implemented during the migration development so I think I left as it was. We can have some improvements but more about if-else on the node role (broker vs controller). For each of them, the not if-else approach is needed because depending on the migration state you could have ZooKeeper connection together with migration flag enabled but not KRaft, or any other combination. Writing a if-else, which depending on each state duplicate a lot of code about what applying would make the code longer and very complex. This way, each if condition defines when a specific configuration (ZK, ZK migration flag or KRaft) need to be applied. It's just a representation of the table I made as comment.
* the `buildBrokerConfiguration` (which I could rename as you suggested in a previous comment) is called via `generatePerBrokerConfiguration`. This is the logic we have today where `buildBrokerConfiguration` is missing just because the logic to build the broker configuration is simpler. This method is just applying the builder pattern on an already existing builder as it stands today.
* Yes yes, any INFO level log will be re-evaluated. They are all INFO, more or less, just for the purpose of a better output on testing.

Reply:
I could change the code this way ...

```java
if ((node.broker() && this.kafkaMetadataConfigState.isZooKeeperOrMigration()) ||
                (node.controller() && this.kafkaMetadataConfigState.isPreMigrationOrKRaft() && this.kafkaMetadataConfigState.isZooKeeperOrPostMigration())) {
            builder.withZookeeper(cluster);
            LOGGER.infoCr(reconciliation, "withZookeeper on node [{}]", node.podName());
        }

        if ((node.broker() && this.kafkaMetadataConfigState.isMigration()) ||
                (node.controller() && this.kafkaMetadataConfigState.isPreMigrationOrKRaft() && this.kafkaMetadataConfigState.isZooKeeperOrPostMigration())) {
            builder.withZooKeeperMigration(true);
            LOGGER.infoCr(reconciliation, "withZooKeeperMigration on node [{}]", node.podName());
        }

        if ((node.broker() && this.kafkaMetadataConfigState.isMigrationOrKRaft()) ||
                (node.controller() && this.kafkaMetadataConfigState.isPreMigrationOrKRaft())) {
            builder.withKRaft(cluster, namespace, pool.processRoles, nodes());
            LOGGER.infoCr(reconciliation, "withKRaft on node [{}]", node.podName());
        }
```

But not sure how much it's better than the current one which is kind of reflecting the table described in the comment.